### ISSUE TITLE
Security: OIDC security provider causes an exception when using a proxy

### ISSUE THREAD

**Issue by SergejMunda** [oracle/helidon#3719]
## Environment Details
* Helidon Version: 2.4.0
* Helidon MP
* JDK version: 17
* OS: Linux (Debian)
* Docker version (if applicable): N/A

----------

## Problem Description

When configuring a OIDC security provider with proxy options (proxy-protocol, proxy-host, proxy-port), the server stops by throwing the following execption (the IDP service uri was replaced with the text "<IDENTITY_PROVIDER_URI>"):

```
Exception in thread "main" java.util.concurrent.CompletionException: io.helidon.webclient.WebClientException <IDENTITY_PROVIDER_URI>
	at java.base/java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:368)
	at java.base/java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:377)
	at java.base/java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:1097)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162)
	at io.helidon.webclient.WebClientRequestBuilderImpl.lambda$invoke$20(WebClientRequestBuilderImpl.java:605)
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:578)
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:552)
	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:491)
	at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:616)
	at io.netty.util.concurrent.DefaultPromise.setFailure0(DefaultPromise.java:609)
	at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:117)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.connect(AbstractNioChannel.java:284)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.connect(DefaultChannelPipeline.java:1342)
	at io.netty.channel.AbstractChannelHandlerContext.invokeConnect(AbstractChannelHandlerContext.java:548)
	at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:533)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.connect(CombinedChannelDuplexHandler.java:495)
	at io.netty.channel.ChannelOutboundHandlerAdapter.connect(ChannelOutboundHandlerAdapter.java:51)
	at io.netty.channel.CombinedChannelDuplexHandler.connect(CombinedChannelDuplexHandler.java:296)
	at io.netty.handler.proxy.HttpProxyHandler$HttpClientCodecWrapper.connect(HttpProxyHandler.java:299)
	at io.netty.channel.AbstractChannelHandlerContext.invokeConnect(AbstractChannelHandlerContext.java:548)
	at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:533)
	at io.netty.handler.proxy.ProxyHandler.connect(ProxyHandler.java:179)
	at io.netty.channel.AbstractChannelHandlerContext.invokeConnect(AbstractChannelHandlerContext.java:548)
	at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:533)
	at io.netty.handler.ssl.SslHandler.connect(SslHandler.java:728)
	at io.netty.channel.AbstractChannelHandlerContext.invokeConnect(AbstractChannelHandlerContext.java:548)
	at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:533)
	at io.netty.handler.logging.LoggingHandler.connect(LoggingHandler.java:231)
	at io.netty.channel.AbstractChannelHandlerContext.invokeConnect(AbstractChannelHandlerContext.java:548)
	at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:533)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.connect(CombinedChannelDuplexHandler.java:495)
	at io.netty.channel.ChannelOutboundHandlerAdapter.connect(ChannelOutboundHandlerAdapter.java:51)
	at io.netty.channel.CombinedChannelDuplexHandler.connect(CombinedChannelDuplexHandler.java:296)
	at io.netty.channel.AbstractChannelHandlerContext.invokeConnect(AbstractChannelHandlerContext.java:548)
	at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:533)
	at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:517)
	at io.netty.channel.DefaultChannelPipeline.connect(DefaultChannelPipeline.java:978)
	at io.netty.channel.AbstractChannel.connect(AbstractChannel.java:265)
	at io.netty.bootstrap.Bootstrap$3.run(Bootstrap.java:250)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: io.helidon.webclient.WebClientException: <IDENTITY_PROVIDER_URI>
	... 41 more
Caused by: java.nio.channels.UnresolvedAddressException
	at java.base/sun.nio.ch.Net.checkAddress(Net.java:149)
	at java.base/sun.nio.ch.Net.checkAddress(Net.java:157)
	at java.base/sun.nio.ch.SocketChannelImpl.checkRemote(SocketChannelImpl.java:816)
	at java.base/sun.nio.ch.SocketChannelImpl.connect(SocketChannelImpl.java:839)
	at io.netty.util.internal.SocketUtils$3.run(SocketUtils.java:91)
	at io.netty.util.internal.SocketUtils$3.run(SocketUtils.java:88)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:569)
	at io.netty.util.internal.SocketUtils.connect(SocketUtils.java:88)
	at io.netty.channel.socket.nio.NioSocketChannel.doConnect(NioSocketChannel.java:315)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.connect(AbstractNioChannel.java:248)
	... 33 more
```

The reason for the exception is an incorrect setup of the io.helidon.webclient.Proxy instance inside the io.helidon.security.providers.oidc.common.OidcUtil.webClientBaseBuilder method:
```java
static WebClient.Builder webClientBaseBuilder(String proxyProtocol,
                                                  String proxyHost,
                                                  int proxyPort,
                                                  Duration clientTimeout) {
        WebClient.Builder webClientBuilder = WebClient.builder()
                .addService(WebClientTracing.create())
                .addMediaSupport(JsonpSupport.create())
                .connectTimeout(clientTimeout.toMillis(), TimeUnit.MILLISECONDS)
                .readTimeout(clientTimeout.toMillis(), TimeUnit.MILLISECONDS);

        if (proxyHost != null) {
            webClientBuilder.proxy(Proxy.builder()
                                           .type(Proxy.ProxyType.HTTP)
                                           .host(proxyProtocol + "://" + proxyHost)
                                           .port(proxyPort)
                                           .build());
        }
        return webClientBuilder;
    }
```

The line `.host(proxyProtocol + "://" + proxyHost)` creates a host name that the java.net.InetSocketAddress constructor cannot resolve because of the added proxyProtocol. This creates an unresolved address:
```java
public InetSocketAddress(String hostname, int port) {
        checkHost(hostname);
        InetAddress addr = null;
        String host = null;
        try {
            addr = InetAddress.getByName(hostname);
        } catch(UnknownHostException e) {
            host = hostname;
        }
        holder = new InetSocketAddressHolder(host, addr, checkPort(port));
    }
```

Before making a request to the IDP service the sun.nio.ch.Net.checkAddress method is called and it throws the above `UnresolvedAddressException` exception:
```java
public static InetSocketAddress checkAddress(SocketAddress sa) {
        if (sa == null)
            throw new NullPointerException();
        if (!(sa instanceof InetSocketAddress))
            throw new UnsupportedAddressTypeException(); // ## needs arg
        InetSocketAddress isa = (InetSocketAddress)sa;
        if (isa.isUnresolved())
            throw new UnresolvedAddressException(); // ## needs arg
        InetAddress addr = isa.getAddress();
        if (!(addr instanceof Inet4Address || addr instanceof Inet6Address))
            throw new IllegalArgumentException("Invalid address type");
        return isa;
    }
```

There is also a inconsistency when retrieving information from the IDP service. the metadata ("/.well-known" endpoint) gets called by the HttpsURLConnection class while the other endpoints get called with the io.helidon.webclient.WebClient class (Netty).

## Steps to reproduce
1. Set the security.providers.X.oidc.proxy-protocol, security.providers.X.oidc.proxy-host and security.providers.X.oidc.proxy-port options to use a proxy server (and the rest of the required OIDC security provider options).
3. Build and run the project, it will throw the exception when trying to access the IDP service through the configured proxy.

**Comment by tomas-langer** [comment]
The inconsistency in clients:
we use a class `Resource` from our common module which cannot depend on `WebClient` (as `WebClient` uses it as well).
The resource can be loaded from multiple locations (classpath, path, HTTP). 
In this case, we use the URL method to get an input stream on the resource.

When we invoke OIDC requests that require more configuration (headers, parameters etc.) we use the `WebClient`.

**Comment by tomas-langer** [comment]
Proxy can be defined for both.

**Comment by barchetta** [comment]
This has been fixed in 2.4.1.

**Comment by SergejMunda** [comment]
> The inconsistency in clients:
we use a class Resource from our common module which cannot depend on WebClient (as WebClient uses it as well).
The resource can be loaded from multiple locations (classpath, path, HTTP).
In this case, we use the URL method to get an input stream on the resource.
When we invoke OIDC requests that require more configuration (headers, parameters etc.) we use the WebClient.

Got it. Thank you for the explanation.

**Comment by SergejMunda** [comment]
> This has been fixed in 2.4.1.

Great. Closing the issue.
